packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -357,18 +361,22 @@ function updateMemoComponent(
       // SimpleMemoComponent codepath doesn't resolve outer props either.
       Component.defaultProps === undefined
     ) {
+      let resolvedType = type;
+      if (__DEV__) {
+        resolvedType = resolveFunctionForHotReloading(type);
```

Comment:
Does this works for updates against the SimpleMemoComponent? Does the thing in ReactFiber know to deal with unwrapping the "Memo" HoC and update the `.type`?

Reply:
My understanding is:

1. First time we "upgrade" to simple memo, type gets resolved on this line.
2. Next time we work on it, `createWorkInProgress` resolves it for simple memo.

The thing in ReactFiber doesn't deal with regular memo at all because we only need resolving for simple memo. Regular memo doesn't call into user code (except the comparison function, but I don't aim to reload it).

Reply:
Can you explain how you expect the data structures for memo to work? There are two. The render function and the object around it. createWorkInProgress will resolve the memo object around it but that's wrong because the type of the `.type` field has now changed and it needs the render function now.

Reply:
My understanding is that normal Memo never runs rendering function by itself. It always either delegates to child fiber, or gets upgraded:

* Either it's Memo-that-does-a-bailout + FunctionComponent (or whatever's in memo).
* Or it gets upgraded to SimpleMemo.

So in practice only FunctionComponent and SimpleMemo need to resolve the function. Conveniently, that's what their type is (a function).

Memo doesn't need to resolve a function because it doesn't run it.

(ForwardRef is an exception because it also runs user code *and* its type is a wrapper object. That's why ForwardRef has a special codepath where we recreate the wrapper object if we only have a new inner function alone.)

The wrapper object doesn't play a role at all, except at the point of `__register__` call (which will does another `__register__` for the inner function).